### PR TITLE
helm: Make DNS policy for agent and operator configurable

### LINKED ISSUE #20079: CFP: [Helm chart] Support configuring cilium agent Pod dnsPolicy
## Cilium Feature Proposal

**Is your feature request related to a problem?**

When using etcd-operator managed etcd as cilium kvstore backend, as documented https://docs.cilium.io/en/v1.9/concepts/kubernetes/requirements/#kube-dns one of the options is to configure `dnsPolicy` for the cilium DaemonSet Pod template. This is not currently configurable in cilium helm chart (OSS and enterprise), and chart alone can't be used to manage cilium installations, have to use extra tools like kustomize to template dnsPolicy.

**Describe the feature you'd like**

Please support configuring optional dnsPolicy for cilium agent in cilium Helm chart.

**(Optional) Describe your proposed solution**

Default to nothing set (Pod default) like now. Use config option when value is provided.

### PR DESCRIPTION
**Pull request by michi-covalent**
Fixes #20079

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

```release-note
helm: Make DNS policy for cilium-agent and cilium-operator pods configurable
```

### CODE REVIEW

**Comment by michi-covalent** [comment]
tested with:
```
helm install -n kube-system cilium ./install/kubernetes/cilium --set dnsPolicy=ClusterFirstWithHostNet --set operator.dnsPolicy=ClusterFirstWithHostNet
```
and verified both agent and operator comes up with `ClusterFirstWithHostNet` dns policy.

File: install/kubernetes/cilium/values.yaml [hunk 1]
```diff
@@ -123,6 +123,10 @@ hostAliases: []
 # -- The priority class to use for cilium-agent.
 priorityClassName: ""
 
+# -- DNS policy for Cilium agent pods.
+# Ref: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy
+dnsPolicy: ""
```
Comment (joestringer) [inline install/kubernetes/cilium/values.yaml:125]: After #20066 was merged, now these changes need to be made to `install/kubernetes/cilium/values.yaml.tmpl` instead, then run `make -C install/kubernetes` to update this file.

Comment (michi-covalent) [reply install/kubernetes/cilium/values.yaml:125]: oh yeah i forgot about this pr 😹 let me update it.

**Review by qmonnet — APPROVED** [review]

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 2]
```diff
@@ -582,6 +582,8 @@ spec:
       # In managed etcd mode, Cilium must be able to resolve the DNS name of
       # the etcd service
       dnsPolicy: ClusterFirstWithHostNet
+      {{- else if .Values.dnsPolicy }}
```
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:629]: Should we document that `.Values.dnsPolicy` is ignored if `Values.etcd.managed` is set?

**Review by gandro — APPROVED** [review]

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml (hunk 2 above)
Comment (michi-covalent) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:629]: updated!

**Review by joestringer — APPROVED** [review]

File: install/kubernetes/cilium/templates/cilium-operator/deployment.yaml [hunk 3]
```diff
@@ -207,6 +207,8 @@ spec:
       # In managed etcd mode, Cilium must be able to resolve the DNS name of
       # the etcd service
       dnsPolicy: ClusterFirstWithHostNet
+      {{- else if .Values.dnsPolicy }}
```
Comment (stevo-f3) [inline install/kubernetes/cilium/templates/cilium-operator/deployment.yaml:210]:
```suggestion
      {{- else if .Values.operator.dnsPolicy }}
```

Comment (michi-covalent) [reply install/kubernetes/cilium/templates/cilium-operator/deployment.yaml:210]: ah good catch 🤦 updated!

**Review by sayboras — APPROVED** [review]
LGMT :100:

**Review by nathanjsweet — APPROVED** [review]

**Comment by michi-covalent** [comment]
added an extra commit to update values.yaml.tmpl looks like it got out of sync: https://github.com/cilium/cilium/pull/20082/commits/9da0702218aa8e121d7356be750c708451fc45de

**Comment by sayboras** [comment]
/test-1.16-4.9

**Comment by nbusseneau** [comment]
@michi-covalent This one cannot be cleanly backported to `v1.11` because the changes to `install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml` depend on changes from #17243, which was not backported to `v1.11`.

I am removing the backport label, please assess the situation and re-add to both if we should indeed backport.

**Comment by michi-covalent** [comment]
thanks for the heads up @nbusseneau i'll take care of backporting this

**Comment by michi-covalent** [comment]
opened https://github.com/cilium/cilium/pull/20842

**Comment by nbusseneau** [comment]
Yea so actually I hadn't read what #17243 was doing: it's not actually adding anything, just re-organizing stuff. So conflicts can be resolved trivially, my bad. Will add back to main backport PR.

**Pull request merged**